java/fury-core/src/main/java/org/apache/fury/meta/MetaStringEncoder.java

```diff
@@ -0,0 +1,288 @@
+   *
+   * @param specialChar1 The first special character used in custom encoding.
+   * @param specialChar2 The second special character used in custom encoding.
+   */
+  public MetaStringEncoder(char specialChar1, char specialChar2) {
+    this.specialChar1 = specialChar1;
+    this.specialChar2 = specialChar2;
+  }
+
+  /**
+   * Encodes the input string to MetaString using adaptive encoding, which intelligently chooses the
+   * best encoding based on the string's content.
+   *
+   * @param input The string to encode.
+   * @return A MetaString object representing the encoded string.
+   */
+  public MetaString encode(String input) {
+    if (input.isEmpty()) {
+      return new MetaString(
+          input, Encoding.LOWER_SPECIAL, specialChar1, specialChar2, new byte[0], 0);
+    }
+    Encoding encoding = computeEncoding(input);
+    return encode(input, encoding);
+  }
+
+  /**
+   * Encodes the input string to MetaString using specified encoding.
+   *
+   * @param input The string to encode.
+   * @param encoding The encoding to use.
+   * @return A MetaString object representing the encoded string.
+   */
+  public MetaString encode(String input, Encoding encoding) {
+    if (input.isEmpty()) {
+      return new MetaString(
+          input, Encoding.LOWER_SPECIAL, specialChar1, specialChar2, new byte[0], 0);
+    }
+    int length = input.length();
+    switch (encoding) {
+      case LOWER_SPECIAL:
+        return new MetaString(
+            input, encoding, specialChar1, specialChar2, encodeLowerSpecial(input), length * 5);
+      case LOWER_UPPER_DIGIT_SPECIAL:
+        return new MetaString(
+            input,
+            encoding,
+            specialChar1,
+            specialChar2,
+            encodeLowerUpperDigitSpecial(input),
+            length * 6);
+      case FIRST_TO_LOWER_SPECIAL:
+        return new MetaString(
+            input,
+            encoding,
+            specialChar1,
+            specialChar2,
+            encodeFirstToLowerSpecial(input),
+            length * 5);
+      case ALL_TO_LOWER_SPECIAL:
+        char[] chars = input.toCharArray();
+        int upperCount = countUppers(chars);
+        return new MetaString(
+            input,
+            encoding,
+            specialChar1,
+            specialChar2,
+            encodeAllToLowerSpecial(chars, upperCount),
+            (upperCount + length) * 5);
+      default:
+        byte[] bytes = input.getBytes(StandardCharsets.UTF_8);
+        return new MetaString(
+            input, Encoding.UTF_8, specialChar1, specialChar2, bytes, bytes.length * 8);
+    }
+  }
+
+  public Encoding computeEncoding(String input) {
+    if (input.isEmpty()) {
+      return Encoding.LOWER_SPECIAL;
+    }
+    char[] chars = input.toCharArray();
+    if (canBeLowerSpecialEncoded(chars)) {
+      return Encoding.LOWER_SPECIAL;
+    } else if (canBeLowerUpperDigitSpecialEncoded(chars)) {
```

Comment:
`canBeLowerUpperDigitSpecialEncoded`, `countDigits` and `countUppers` traverse `chars` three times. Could we traverse the `chars` only once in `canBeLowerUpperDigitSpecialEncoded`, calculate the digits of numbers, the number of uppercase letters and return a bean(the number of characters of both types that should be included in the bean).

Reply:
updated